File: infra/src/main/java/com/beat/infra/persistence/booking/mapper/BookingPersistenceMapper.java
package com.beat.infra.persistence.booking.mapper;

import org.springframework.stereotype.Component;

import com.beat.domain.booking.model.Booking;
import com.beat.domain.booking.vo.RefundAccount;
import com.beat.domain.exception.DomainException;
import com.beat.infra.persistence.booking.entity.BookingJpaEntity;
import com.beat.infra.persistence.booking.entity.RefundAccountJpaValue;
import com.beat.infra.persistence.exception.PersistenceMappingException;

@Component
public class BookingPersistenceMapper {

	public Booking toDomain(BookingJpaEntity entity) {
		try {
			return Booking.rehydrate(
				entity.getId(),
				entity.getPurchaseTicketCount(),
				entity.getBookerName(),
				entity.getBookerPhoneNumber(),
				entity.getBookingStatus(),
				entity.getCreatedAt(),
				entity.getCancellationDate(),
				entity.getBirthDate(),
				entity.getPassword(),
				toDomain(entity.getRefundAccount()),
				entity.getScheduleId(),
				entity.getUserId(),
				entity.getTotalPaymentAmount()
			);
		} catch (DomainException exception) {
			throw PersistenceMappingException.invalidStoredState("Booking", entity.getId(), exception);
		}
	}

	public BookingJpaEntity toEntity(Booking domain) {
		return BookingJpaEntity.rehydrate(
			domain.getId(),
			domain.getPurchaseTicketCount(),
			domain.getBookerName(),
			domain.getBookerPhoneNumber(),
			domain.getBookingStatus(),
			domain.getCreatedAt(),
			domain.getCancellationDate(),
			domain.getBirthDate(),
			domain.getPassword(),
			toEntity(domain.getRefundAccount()),
			domain.getScheduleId(),
			domain.getUserId(),
			domain.getTotalPaymentAmount()
		);
	}

	private RefundAccount toDomain(RefundAccountJpaValue value) {
		if (value == null) {
			return null;
		}
		return RefundAccount.of(value.getBankName(), value.getAccountNumber(), value.getAccountHolder());
	}

	private RefundAccountJpaValue toEntity(RefundAccount value) {
		if (value == null) {
			return null;
		}
		return new RefundAccountJpaValue(value.getBankName(), value.getAccountNumber(), value.getAccountHolder());
	}
}

File: core/infra/src/main/kotlin/com/beat/infra/persistence/booking/mapper/BookingPersistenceMapper.kt
package com.beat.infra.persistence.booking.mapper

import com.beat.domain.booking.model.Booking
import com.beat.domain.booking.vo.RefundAccount
import com.beat.domain.exception.DomainException
import com.beat.infra.persistence.booking.entity.BookingJpaEntity
import com.beat.infra.persistence.booking.entity.RefundAccountJpaValue
import com.beat.infra.persistence.exception.PersistenceMappingException
import org.springframework.stereotype.Component

@Component
class BookingPersistenceMapper {
    fun toDomain(entity: BookingJpaEntity): Booking =
        try {
            Booking.rehydrate(
                entity.id,
                entity.purchaseTicketCount,
                entity.bookerName,
                entity.bookerPhoneNumber,
                entity.bookingStatus,
                entity.createdAt,
                entity.cancellationDate,
                entity.birthDate,
                entity.password,
                toDomain(entity.refundAccount),
                entity.scheduleId,
                entity.userId,
                entity.totalPaymentAmount,
            )
        } catch (exception: DomainException) {
            throw PersistenceMappingException.invalidStoredState("Booking", entity.id, exception)
        }

    fun toEntity(domain: Booking): BookingJpaEntity =
        BookingJpaEntity.rehydrate(
            domain.getId(),
            domain.getPurchaseTicketCount(),
            domain.getBookerName(),
            domain.getBookerPhoneNumber(),
            domain.getBookingStatus(),
            domain.getCreatedAt(),
            domain.getCancellationDate(),
            domain.getBirthDate(),
            domain.getPassword(),
            toEntity(domain.getRefundAccount()),
            domain.getScheduleId(),
            domain.getUserId(),
            domain.getTotalPaymentAmount(),
        )

    private fun toDomain(value: RefundAccountJpaValue?): RefundAccount? =
        value?.let { RefundAccount.of(it.bankName, it.accountNumber, it.accountHolder) }

    private fun toEntity(value: RefundAccount?): RefundAccountJpaValue? =
        value?.let { RefundAccountJpaValue(it.bankName, it.accountNumber, it.accountHolder) }
}
